drivers: adc: add max1125x driver

mustafaabdullahk opened the pull request:
Hi everyone !

This adds support for the max1125x (max11254, max11254) family of spi adc devices.

Signed-off-by: Mustafa Abdullah Kus <mustafa.kus@sparsetechnology.com>

mustafaabdullahk commented:
@nashif @galak @anangl ping

galak commented on dts/bindings/adc/maxim,max1125x-base.yaml:
```diff
@@ -0,0 +1,40 @@
+# Common fields for Texas Instruments MAX1125x
```
Is Texas Instruments correct here?

mustafaabdullahk replied:
No, incorrect. i will replace with 'Analog Devices Inc.'

MaureenHelm reviewed (changes requested):
Thank you for your contribution!

mustafaabdullahk commented on drivers/adc/adc_max1125x.c:
```diff
@@ -0,0 +1,871 @@
+	const struct spi_buf rx_buf = {.buf = buffer, .len = reg_size};
+	const struct spi_buf_set rx = {.buffers = &rx_buf, .count = 1};
+
+	gpio_pin_set_dt(&config->csb_gpio, 1);
```
I know, I'm tried to implement it with ''struct spi_cs_control'' but the chip is not compatible with spi timing requirements.

MaureenHelm replied:
Which timing requirement exactly? Is this perhaps an issue with the SPI driver not supporting CS delay? Which SPI driver are you using?

mustafaabdullahk replied:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/018e7154-69e0-4e3a-928b-8ecd379714fb)

i use stm32 spi driver. i see another adc drivers communicating with same way. for e.g

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/56f9c364-984a-44a3-ae51-7e9f5a08cee0)
adc_ads114s0x.c source file

mustafaabdullahk commented on drivers/adc/adc_max1125x.c:
```diff
@@ -0,0 +1,871 @@
+	case MAX1125X_CONFIG_RATE_125:
+		odr = MAX1125X_CONFIG_RATE_125;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_125];
+		break;
+	case MAX1125X_CONFIG_RATE_250:
+		odr = MAX1125X_CONFIG_RATE_250;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_250];
+		break;
+	case MAX1125X_CONFIG_RATE_500:
+		odr = MAX1125X_CONFIG_RATE_500;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_500];
+		break;
+	case MAX1125X_CONFIG_RATE_1000:
+		odr = MAX1125X_CONFIG_RATE_1000;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_1000];
+		break;
+	case MAX1125X_CONFIG_RATE_2000:
+		odr = MAX1125X_CONFIG_RATE_2000;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_2000];
+		break;
+	case MAX1125X_CONFIG_RATE_4000:
+		odr = MAX1125X_CONFIG_RATE_4000;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_4000];
+		break;
+	case MAX1125X_CONFIG_RATE_8000:
+		odr = MAX1125X_CONFIG_RATE_8000;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_8000];
+		break;
+	case MAX1125X_CONFIG_RATE_16000:
+		odr = MAX1125X_CONFIG_RATE_16000;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_16000];
+		break;
+	case MAX1125X_CONFIG_RATE_32000:
+		odr = MAX1125X_CONFIG_RATE_32000;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_32000];
+		break;
+	case MAX1125X_CONFIG_RATE_64000:
+		odr = MAX1125X_CONFIG_RATE_64000;
+		odr_delay_us = odr_delay[MAX1125X_CONFIG_RATE_64000];
+		break;
```
yep, you are right. i will refactor this part.

mustafaabdullahk commented:
> Thank you for your contribution!



> MaureenHelm

:star_struck:

MaureenHelm commented on drivers/adc/adc_max1125x.c:
```diff
@@ -0,0 +1,811 @@
+
+	k_sleep(data->ready_time);
+
+	uint8_t is_ready
```
```suggestion
	uint8_t is_ready;
```
This leads me to believe this driver isn't getting built in CI. Have you tried building `tests/drivers/build_all/adc` locally?

mustafaabdullahk replied:
yes i run tested before pushed.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/92085323-f436-4a95-9f5c-06d705a0b144)

MaureenHelm commented on dts/bindings/adc/maxim,max1125x-base.yaml:
```diff
@@ -0,0 +1,42 @@
+  gpio1-enable:
+    type: boolean
+    description: this option for gpio1 enable or disable
+  gpio0-direction:
+    type: int
+    enum:
+      - 0
+      - 1
+    description: this option for gpio0 direction 1 for input, 0 for output
+  gpio1-direction:
+    type: int
+    enum:
+      - 0
+      - 1
+    description: this option for gpio0 direction 1 for input, 0 for output
```
These should be booleans

mustafaabdullahk replied:
update as boolean

mustafaabdullahk committed:
drivers: adc: add max1125x driver

This adds support for the max1125x (max11254, max11254)
family of spi adc devices.

Signed-off-by: Mustafa Abdullah Kus <mustafa.kus@sparsetechnology.com>
